src/import.c

```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "storage.h"
+#include "util.h"
+#include "module.h"
+#include "import.h"
+#include "entry.h"
+#include <string.h>
+#include <stdio.h>
+
+
+/**
+ * CacheImport:
+ * @shadow_path: Path to shadow file.
+ * @shadow_hash: Hashed password. 
+ * @storage_path: Path to store policy cache entry.
+ * @username: Username to create cache entry for.
+ * @error: Errors.
+ *
+ * Returns: #True if CacheEntry was created or #False on failure.
+ */
+
+gboolean CacheImport(const char *shadow_path, const char *storage_path, const char *username, GError **error) {
+  gboolean result = TRUE;
+
+  if (!shadow_path) {
+    shadow_path = "/etc/shadow";
+  }
+  
+  if (!storage_path) {
+    storage_path = DEFAULT_STORAGE_PATH;
+  }
+
+  GError *errors = NULL;
+  // Get user's hashed password.
+  GBytes* shadow_hash = CacheUtilReadShadowFile(shadow_path, username, &errors);
```

Comment:
If the result is NULL, you'll want to skip to the error handling. If not, it's going to try and use a NULL shadow_hash which looks like it may be the source of one of the errors.

GLib error handling using GError is almost always actually two things:

1. A GError* that may be NULL for details about the error. If the caller doesn't care about details, they can just pass NULL.
2. A return value that can be the authoritative source of whether or not the call failed. It's often a gboolean that's false on failure or a pointer type that's NULL on failure. Callers always use that to see if they need to jump into error handling logic, so most calls to a function that takes GError* will be surrounded by or immediately followed by an if statement with error handling logic like "goto done" to skip to the bottom of the function.